### PR TITLE
rbac: implement inventory rbac access permissions.

### PR DESCRIPTION
**Pull request by ldjebran**
# Description
- Fixed deviceview api add inventory group fields
- Added Configuration for rbac base url and timeout
- Added feature flag edgeParity.inventory-rba.
- Added github.com/RedHatInsights/rbac-client-go dependency.
- Created rbac client and unit-tests.
- Updated deviceview GET/POST to handle rbac access checking/filters and covered by unittests.

Notes:  

- Use same rules as inventory 
https://github.com/RedHatInsights/insights-host-inventory/blob/master/lib/middleware.py#L75
https://github.com/RedHatInsights/insights-host-inventory/blob/master/api/filtering/filtering.py#L370

- For the feature to be used both feature flags must be enabled 
     - edgeParity.inventory-rbac 
     - edgeParity.inventory-groups-enabled
- This feature is disabled for orgs that are enforced to use edge groups .

FIXES: https://issues.redhat.com/browse/THEEDGE-3624

## Type of change

What is it?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [ ] Refactor

# Checklist:

- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I run `make pre-commit` to check fmt/vet/lint/test-no-fdo

### CODE REVIEW

File: pkg/clients/rbac/client.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+const DefaultTimeDuration = 1 * time.Second
+
+// ClientInterface is an Interface to make request to insights rbac
+type ClientInterface interface {
+	GetAccessList(application Application) (rbacClient.AccessList, error)
+	GetInventoryGroupsAccess(acl rbacClient.AccessList, resource ResourceType, accessType AccessType) (bool, []string, bool, error)
+}
+
+type WrappedClientInterface interface {
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:53]: Docstring please

File: pkg/clients/rbac/client.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+// InitClient initializes the client for Rbac service
+func InitClient(ctx context.Context, log *log.Entry) ClientInterface {
+	return &Client{ctx: ctx, log: log.WithField("client-context", "rbac-client")}
+}
+
+func (c *Client) NewRbacClient(application Application) (WrappedClientInterface, error) {
+	url, err := url2.JoinPath(config.Get().RbacBaseURL, APIPath)
+	if err != nil {
+		c.log.WithField("error", err.Error()).Error("error occurred while creating rbac url")
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:71]: Could you use `ErrCreatingRbacURL`?

File: pkg/clients/rbac/client.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	ctx context.Context
+	log *log.Entry
+}
+
+// InitClient initializes the client for Rbac service
+func InitClient(ctx context.Context, log *log.Entry) ClientInterface {
+	return &Client{ctx: ctx, log: log.WithField("client-context", "rbac-client")}
+}
+
+func (c *Client) NewRbacClient(application Application) (WrappedClientInterface, error) {
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:68]: Docstring

File: pkg/clients/rbac/client.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+		c.log.WithField("error", err.Error()).Error("error occurred while creating rbac url")
+		return nil, ErrCreatingRbacURL
+	}
+
+	wrappedClient := rbacClient.NewClient(url, string(application))
+	wrappedClient.HTTPClient = clients.ConfigureClientWithTLS(wrappedClient.HTTPClient)
+	return &wrappedClient, nil
+}
+
+func (c *Client) GetAccessList(application Application) (rbacClient.AccessList, error) {
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:80]: Docstring

File: pkg/clients/rbac/client.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	conf := config.Get()
+	rbacTimeout := time.Duration(conf.RbacTimeout) * DefaultTimeDuration
+
+	ctx, cancel := context.WithTimeout(c.ctx, rbacTimeout)
+	defer cancel()
+
+	wrappedClient, err := c.NewRbacClient(application)
+	if err != nil {
+		c.log.WithField("error", err.Error()).Error("error occurred when creating rbac client")
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:89]: Maybe turn error into constant?

File: pkg/clients/rbac/client.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+		return nil, err
+	}
+
+	var identity string
+	if config.Get().Auth {
+		identity, err = common.GetOriginalIdentity(ctx)
+		if err != nil {
+			c.log.WithField("error", err.Error()).Error("error getting identity from context")
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:97]: Error as constant?

File: pkg/clients/rbac/client.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+			return nil, ErrGettingIdentityFromContext
+		}
+	}
+
+	acl, err := wrappedClient.GetAccess(ctx, identity, "")
+	if err != nil {
+		c.log.WithField("error", err.Error()).Error("error occurred getting rbac AccessList")
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:104]: There's an error already that you might be able to use?

File: pkg/clients/rbac/client.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+		return nil, err
+	}
+	return acl, nil
+}
+
+func (c *Client) GetInventoryGroupsAccess(acl rbacClient.AccessList, resource ResourceType, accessType AccessType) (bool, []string, bool, error) {
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:110]: Docstring

File: pkg/clients/rbac/client.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+	var overallGroupIDS []string
+	var overallGroupIDSMap = make(map[string]bool)
+	var allowedAccess bool
+	var hostsWithNoGroupsAssigned bool
+	for _, ac := range acl {
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:115]: This part of the code looks pretty convoluted with lots of conditionals and checks. Could it be refactored into separate functions for clarity?

File: pkg/clients/rbac/client.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+						}
+					}
+				}
+			}
+		}
+	}
+	return allowedAccess, overallGroupIDS, hostsWithNoGroupsAssigned, nil
+}
+
+func AccessMatch(access1, access2 AccessType) bool {
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:152]: Docstring

File: pkg/clients/rbac/client.go [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+	return access1 == access2 || access1 == AccessTypeAny
+}
+
+func ResourceMatch(resource1, resource2 ResourceType) bool {
```
Comment (omaciel) [inline pkg/clients/rbac/client.go:156]: Docstring

File: pkg/errors/api.go [hunk 12]
```diff
@@ -86,3 +86,43 @@ func NewFeatureNotAvailable(message string) APIError {
...
+// ServiceUnavailableDefaultMessage the message used by default for Service Unavailable API error
+const ServiceUnavailableDefaultMessage = "service is unavailable"
+
+// NewServiceUnavailable creates a new ServiceUnavailable API error
+func NewServiceUnavailable(message string) APIError {
+	if message == "" {
+		message = ServiceUnavailableDefaultMessage
+	}
+	err := new(FeatureNotAvailable)
+	err.Code = "FORBIDDEN"
```
Comment (omaciel) [inline pkg/errors/api.go:124]: Should the code say forbidden or unavailable?

File: pkg/routes/devices.go [hunk 13]
```diff
@@ -336,6 +338,50 @@ func GetDeviceDBInfo(w http.ResponseWriter, r *http.Request) {
...
+	if identity, err := common.GetIdentityFromContext(r.Context()); err != nil {
+		contextServices.Log.WithField("error", err.Error()).Error("error occurred when retrieving identity from context")
+		respondWithAPIError(w, contextServices.Log, errors.NewBadRequest("error retrieving identity"))
+		return nil, err
+	} else if identity.Identity.Type != common.IdentityTypeUser {
+		return nil, nil
+	}
+	acl, err := contextServices.RbacService.GetAccessList(rbac.ApplicationInventory)
+	if err != nil {
+		contextServices.Log.WithField("error", err.Error()).Error("error occurred when getting rbac access list")
```
Comment (omaciel) [inline pkg/routes/devices.go:352]: Can this error message and others above be reused from existing constants?

File: pkg/routes/devices.go [hunk 14]
```diff
@@ -336,6 +338,50 @@ func GetDeviceDBInfo(w http.ResponseWriter, r *http.Request) {
...
+		return nil, err
+	}
+	if !allowedAccess {
+		apiError := errors.NewForbidden("access to hosts is forbidden")
+		respondWithAPIError(w, contextServices.Log, apiError)
+		return nil, apiError
+	}
+	var filter *gorm.DB
+	if len(groupIDS) > 0 {
+		filter = db.DB.Where("devices.group_uuid IN (?)", groupIDS)
```
Comment (omaciel) [inline pkg/routes/devices.go:372]: Can you add a comment explaining the logic? 6 months from now we might not remember the why :)

File: pkg/routes/devices.go [hunk 15]
```diff
@@ -336,6 +338,50 @@ func GetDeviceDBInfo(w http.ResponseWriter, r *http.Request) {
...
+	}
+	if hostsWithNoGroupsAssigned {
+		noGroupsFilters := db.DB.Where("devices.group_uuid = '' OR devices.group_uuid IS NULL")
```
Comment (omaciel) [inline pkg/routes/devices.go:375]: Can you add a comment explaining the logic?

**Review by omaciel — COMMENTED** [review]
I have done a very quick review and added suggestions and questions.

Please don't block on me :)

File: pkg/errors/api.go (hunk 12 above)
Comment (ldjebran) [reply pkg/errors/api.go:124]: good catch , thanks ""UNAVAILABLE"

File: pkg/clients/rbac/client.go (hunk 5 above)
Comment (ldjebran) [reply pkg/clients/rbac/client.go:89]: this is an error log message , used only at this point, is it making sense to create a const only for this message + when we see this message in logs we can directly locate this message  code line

Comment (omaciel) [reply pkg/clients/rbac/client.go:89]: You're right. I thought I had seen other places where this was repeated but I was wrong. Marking it as resolved.

**Comment by ldjebran** [comment]
@omaciel addressed comments, except  for log messages as we discussed offline

File: pkg/clients/rbac/client.go (hunk 9 above)
Comment (omaciel) [reply pkg/clients/rbac/client.go:155]: Looks much cleaner. Can you now document what each section is doing and why?

**Review by omaciel — COMMENTED** [review]
One final nitpick :)

File: pkg/clients/rbac/client.go (hunk 2 above)
Comment (ldjebran) [reply pkg/clients/rbac/client.go]: I added ErrCreatingRbacURL.Error() as message

File: pkg/errors/api.go (hunk 12 above)
Comment (ldjebran) [reply pkg/errors/api.go]: done

**Review by omaciel — APPROVED** [review]
/lgtm

Thank you

**Review by acosferreira — APPROVED** [review]
lgtm

**Pull request merged**